tensilelite/Tensile/KernelWriterAssembly.py

```diff
@@ -1394,169 +1395,174 @@ def calculateWG():
-      extLabel    = Label(label="LoadExternalStruct", comment="")
-      extLabelEnd = Label(label="LoadExternalStructEnd", comment="")
-      if kernel["ProblemType"]["SupportUserArgs"]:
-        module.addComment0("Check if custom structure pointer is null")
-        module.add(SCmpEQU32(src0=sgpr("ArgType"), src1=2, comment="ArgType == 2 ?"))
-        module.add(SCBranchSCC1(labelName=extLabel.getLabelName(), comment="branch if ArgType == 2"))
-      module.addComment1("Grouped Gemm: offset argument address to gemm")
-      module.addComment0("Grouped Gemm: offset address from wg_table_start to args_start")
-      module.add(SLShiftLeft2AddU32(dst=sgpr("KernArgAddress"), src0=sgpr(sgprNumsOfGemm), src1=sgpr("KernArgAddress")))
-      module.add(SAddCU32(dst=sgpr("KernArgAddress+1"), src0=sgpr("KernArgAddress+1"), src1=hex(0)))
-      module.addComment0("Grouped Gemm: offset address from args_start to gemm_start")
-      module.add(SMulI32(dst=sgpr(tmpSgprGemmIdxLeft), src0=sgpr(tmpSgprGemmIdxLeft),\
-                         src1=(self.argLoader.getOffset() + (numStoreSgprToLoad * 4))))
-      module.add(SAddU32(dst=sgpr("KernArgAddress"), src0=sgpr("KernArgAddress"), src1=sgpr(tmpSgprGemmIdxLeft)))
-      module.add(SAddCU32(dst=sgpr("KernArgAddress+1"), src0=sgpr("KernArgAddress+1"), src1=hex(0)))
-      module.add(self.getKernelArgLoadModule(kernel, sgprStart, load, 4))
-      if kernel["ProblemType"]["SupportUserArgs"]:
-        module.add(SBranch(extLabelEnd.getLabelName()))
-        module.add(extLabel)
+        with self.allocTmpSgpr(6, 2) as tmpSgprRes:
+          tmpSgpr      = tmpSgprRes.idx
+          tmpSgprRes1  = RegisterPoolResource(idx=tmpSgprRes.idx, size=3)
+
+          tmpSgpr0 = tmpSgprRes.idx + 3
+          tmpSgpr1 = tmpSgpr0+1
+          tmpSgpr2 = tmpSgpr1+1
+
+          WGMXCC = self.states.kernel["WorkGroupMappingXCC"]
+          label_skipWGMXCC = Label(label="skip_WGMXCC", comment="skip WGMXCC if no enough WGs to remap")
+          module.addComment0("only remap WGs in the range")
+          module.add(scalarStaticDivideAndRemainder(qReg=sgpr(tmpSgpr0), rReg=None, dReg=sgpr(tmpSgprNumWorkGroups), divisor=pow(WGMXCC,2), tmpSgprRes=tmpSgprRes1, doRemainder=0))
+          module.add(SMulI32(dst=sgpr(tmpSgpr0), src0=sgpr(tmpSgpr0), src1=pow(WGMXCC,2)))
+          module.add(SCmpGeU32(src0=sgpr("WorkGroup0"), src1=sgpr(tmpSgpr0)))
+          module.add(SCBranchSCC1(label_skipWGMXCC.getLabelName()))
+          module.addComment0("temp0 = (wg*%u)%%%u"%(WGMXCC,pow(WGMXCC,2)))
+          module.add(SMulI32(dst=sgpr(tmpSgpr0), src0=sgpr("WorkGroup0"), src1=WGMXCC))
+          module.add(scalarStaticRemainder(qReg=tmpSgpr, rReg=tmpSgpr0, dReg=tmpSgpr0, divisor=pow(WGMXCC,2), tmpSgprRes=tmpSgprRes1))
+          module.addComment0("temp1 = (wg%%%u)//%u"%(pow(WGMXCC,2), WGMXCC))
+          module.add(scalarStaticRemainder(qReg=tmpSgpr, rReg=tmpSgpr1, dReg="WorkGroup0", divisor=pow(WGMXCC,2), tmpSgprRes=tmpSgprRes1))
```

Comment:
Not sure whether we can set qReg and tmpSgprRes the same register?

Reply:
yes cause we don't need this output